compress.go

```diff
@@ -74,77 +74,65 @@ func CompressHandlerLevel(h http.Handler, level int) http.Handler {
 		level = gzip.DefaultCompression
 	}
 
+	const (
+		gzipEncoding  = "gzip"
+		flateEncoding = "deflate"
+	)
+
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-	L:
-		for _, enc := range strings.Split(r.Header.Get("Accept-Encoding"), ",") {
-			switch strings.TrimSpace(enc) {
-			case "gzip":
-				w.Header().Set("Content-Encoding", "gzip")
-				r.Header.Del("Accept-Encoding")
-				w.Header().Add("Vary", "Accept-Encoding")
-
-				gw, _ := gzip.NewWriterLevel(w, level)
-				defer gw.Close()
-
-				h, hok := w.(http.Hijacker)
-				if !hok { /* w is not Hijacker... oh well... */
-					h = nil
-				}
-
-				f, fok := w.(http.Flusher)
-				if !fok {
-					f = nil
-				}
-
-				cn, cnok := w.(http.CloseNotifier)
-				if !cnok {
-					cn = nil
-				}
-
-				w = &compressResponseWriter{
-					Writer:         gw,
-					ResponseWriter: w,
-					Hijacker:       h,
-					Flusher:        f,
-					CloseNotifier:  cn,
-				}
-
-				break L
-			case "deflate":
-				w.Header().Set("Content-Encoding", "deflate")
-				r.Header.Del("Accept-Encoding")
-				w.Header().Add("Vary", "Accept-Encoding")
-
-				fw, _ := flate.NewWriter(w, level)
-				defer fw.Close()
-
-				h, hok := w.(http.Hijacker)
-				if !hok { /* w is not Hijacker... oh well... */
-					h = nil
-				}
-
-				f, fok := w.(http.Flusher)
-				if !fok {
-					f = nil
-				}
-
-				cn, cnok := w.(http.CloseNotifier)
-				if !cnok {
-					cn = nil
-				}
-
-				w = &compressResponseWriter{
-					Writer:         fw,
-					ResponseWriter: w,
-					Hijacker:       h,
-					Flusher:        f,
-					CloseNotifier:  cn,
-				}
-
-				break L
+		// detect what encoding to use
+		var encoding string
+		for _, curEnc := range strings.Split(r.Header.Get("Accept-Encoding"), ",") {
```

Comment:
I know we could also setup the `encWriter` here, but I tried that and I think the code is more readable when this loop is only responsible for detecting the encoding.